### PR TITLE
Add library base

### PR DESCRIPTION
**Pull request by ameech**
Provides a base for building console features within your app.

Light on tests, but want to put this up for discussion.

@equip/contributors

### CODE REVIEW

File: src/Application.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    /**
+     * Start the console application
+     * 
+     * @return int 0 if everything went fine, or an error code
+     */
+    public function run()
+    {
+        $this->configuration->apply($this->injector);
+        
+        $application = $this->injector->make(ConsoleApplication::class);
```
Comment (ameech) [inline src/Application.php:98]: I don't love this. Is there a better way to load a configured `Application`?

File: composer.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    "authors": [
+        {
+            "name": "Equip Contributors",
+            "homepage": "https://github.com/equip"
+        }
+    ],
+    "require": {
+        "equip/config": "^1.0",
+        "equip/structure": "^1.1",
+        "php": ">=5.5",
```
Comment (shadowhand) [inline composer.json]: Despite the alphabetical organization, I think this line should always go first.

File: composer.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        "phpunit/phpunit": "^4.8|^5.0"
+    },
+    "autoload": {
+        "psr-4": {
+            "Equip\\Console\\": "src/"
+        }
+    },
+    "autoload-dev": {
+        "psr-4": {
+            "EquipTests\\Console\\": "tests/"
```
Comment (shadowhand) [inline composer.json]: I wouldn't use a separate namespace for this. It's way easier to use the same namespace and not have to import everything.

File: console.php [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<?php
```
Comment (shadowhand) [inline console.php]: Maybe make this file bin/console?

``` php
#!/usr/bin/env php
<?php
```

File: src/Application.php (hunk 1 above)
Comment (shadowhand) [inline src/Application.php:98]: I think it's totally fine.

File: src/Command/CommandSet.php [hunk 5]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace Equip\Console\Command;
+
+use Equip\Structure\Set;
+
+class CommandSet extends Set
+{
+
```
Comment (shadowhand) [inline src/Command/CommandSet.php]: You probably want a validation method in here to ensure the classes added extend the command interface.

File: composer.json (hunk 3 above)
Comment (ameech) [inline composer.json]: Left over from `equip/skeleton`, removed from this repo now.

File: console.php (hunk 4 above)
Comment (ameech) [inline console.php]: Whoops, yep! Fixed.

File: src/Command/CommandSet.php (hunk 5 above)
Comment (ameech) [inline src/Command/CommandSet.php]: Added!

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  - 7.0
+  - hhvm
+
+matrix:
+  include:
+    - php: 5.5
+      env: 'COMPOSER_FLAGS="--prefer-stable --prefer-lowest"'
+
+before_script:
+  - travis_retry composer self-update
```
Comment (elazar) [inline .travis.yml]: This isn't needed anymore. Travis VMs do this by default.

File: LICENSE [hunk 7]
```diff
@@ -0,0 +1,22 @@
+The MIT License (MIT)
+
+Copyright (c) :year :owner <@:twitter>
```
Comment (elazar) [inline LICENSE]: Details.

Comment (ameech) [inline LICENSE]: Yep, we're just updated.

File: src/Application.php [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+     * 
+     * @param array $commands
+     * 
+     * @return $this
+     */
+    public function setCommands(array $commands)
+    {
+        $this->commands = $this->commands->withValues($commands);
+        return $this;
+    }
```
Comment (elazar) [inline src/Application.php:87]: These two accessor methods seem superfluous. If I want to modify either object, I'd probably do it in configuration outside this class.

File: bin/console [hunk 9]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env php
+<?php
```
Comment (ameech) [inline bin/console:2]: Not sure we should even include this file. You won't be able to add new commands unless you implement it yourself.

**Comment by elazar** [comment]
Overall, it seems like this is primarily a library that bootstraps symfony/console using Auryn and equip/config?

**Comment by ameech** [comment]
More or less, yep.

File: bin/console (hunk 9 above)
Comment (elazar) [inline bin/console:2]: It may be better as an example for users to write their own?

Comment (ameech) [inline bin/console:2]: Sure, that makes sense. I'll leave it there as an example.

**Comment by shadowhand** [comment]
:+1:

File: src/Application.php (hunk 8 above)
Comment (ameech) [inline src/Application.php:87]: Would you rather see:

``` PHP
$configuration = new ConfigurationSet;
$configuration = $configuration->withValues([
  ..
]);
Application::build($injector, $configuration, $commands);
```

Personally, I think the current implementation reads quite well and is more or less self documenting.

Comment (elazar) [inline src/Application.php:87]: In the particular case of `ConfigurationSet`, I usually subclass it and handle setting its values in the constructor. Suspect I'd do something similar in `CommandSet`. That way, those sets can be re-used in different contexts.

Comment (ameech) [inline src/Application.php:87]: Which you could do anyways, this just makes it easier to get started.

**Pull request merged**